pkg/scheduler/framework/runtime/framework.go

```diff
@@ -676,15 +676,46 @@ func (f *frameworkImpl) QueueSortFunc() framework.LessFunc {
 // When it returns Skip status, returned PreFilterResult and other fields in status are just ignored,
 // and coupled Filter plugin/PreFilterExtensions() will be skipped in this scheduling cycle.
 // If a non-success status is returned, then the scheduling cycle is aborted.
-func (f *frameworkImpl) RunPreFilterPlugins(ctx context.Context, state *framework.CycleState, pod *v1.Pod) (_ *framework.PreFilterResult, status *framework.Status) {
+func (f *frameworkImpl) RunPreFilterPlugins(ctx context.Context, state *framework.CycleState, pod *v1.Pod, diagnosis *framework.Diagnosis, allNodes []*framework.NodeInfo) (_ *framework.PreFilterResult, status *framework.Status) {
 	startTime := time.Now()
 	skipPlugins := sets.New[string]()
+	message := func(plugins []string) string {
+		if len(plugins) == 0 {
+			return ""
+		}
+		msg := fmt.Sprintf("node(s) didn't satisfy plugin(s) %v simultaneously", plugins)
+		if len(plugins) == 1 {
+			msg = fmt.Sprintf("node(s) didn't satisfy plugin %v", plugins[0])
+		}
+		return msg
+	}
+	preFilterNodeToStatusMap := framework.NodeToStatusMap{}
+	plugins := sets.Set[string]{}
 	defer func() {
 		state.SkipFilterPlugins = skipPlugins
+		var preFilterMsg string
+		for node, pluginStatus := range preFilterNodeToStatusMap {
+			if len(pluginStatus.Reasons()) == 0 {
+				pluginStatus.AppendReason(message(pluginStatus.Plugin()))
+			} else {
+				preFilterMsg = pluginStatus.Message()
+			}
+			plugins.Insert(pluginStatus.Plugin()...)
+			diagnosis.AddPluginStatus(pluginStatus)
+			diagnosis.NodeToStatusMap[node] = pluginStatus
+
+		}
+		if len(preFilterMsg) == 0 {
+			preFilterMsg = message(plugins.UnsortedList())
+		}
+		diagnosis.PreFilterMsg = preFilterMsg
```

Comment:
There's too many things happening in `defer`, especially it looks much harder to understand the logic around `preFilterMsg`. I guess it's simpler to keep [these lines](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/schedule_one.go#L464-L468), rather than trying to do everything of diagnosis in `RunPreFilterPlugins`.